### ISSUE TITLE
Forgery limits on packet protection

### ISSUE THREAD

**Issue by martinthomson** [quicwg/base-drafts#3619]
tl;dr We need to recommend limits on the number of failed decryptions.

## Introduction

At [the QUIPS workshop](https://www.ndss-symposium.org/ndss2020/cfp-quips-workshop/) one of the side discussions concentrated on the quality of the analysis of the [record limits in TLS](https://tools.ietf.org/html/rfc8446#section-5.5) and the applicability of that analysis to QUIC.

Felix Günther, along with Marc Fischlin, Christian Janson, and Kenny Paterson have done a little analysis, based on [the QUIPS paper Felix presented at QUIPS](https://felixguenther.info/Q20_RC.pdf) and the [paper by Atul Luykx and Kenny](https://www.isg.rhul.ac.uk/~kp/TLS-AEbounds.pdf) we used to set limits in TLS.

There are two relevant limits in this analysis:

1. The confidentiality bound, which is usually expressed in terms of advantage an attacker gains in being able to break confidentiality of records/packets as they get more data from valid packets.  For instance, in AES-GCM, the analysis shows that after seeing 2^24.5 packets (almost 24 million) the attacker has a 2^-60 better chance than pure chance of winning the IND-CPA game.

2. The forgery bound, which is again expressed as an probably of an attacker successfully creating a forged record/packet.  Thanks to the analysis done previously, we know that AES-GCM and ChaCha20+Poly1305 have a forgery chance of 2^-57 after 2^60 and 2^36 forgery attempts respectively.

The current text just points to TLS 1.3, which recommends the use of key updates to protect confidentiality.  But TLS has a forgery tolerance of 0: a single failed forgery attempt causes the connection to break.  QUIC (and DTLS) allow for multiple forgeries to be ignored, so we need to factor that in.

## Proposed Solution

As discussed in [this paper](https://felixguenther.info/Q20_RC.pdf) if we want to maintain security, we need to consider both confidentiality and integrity.  For that we need two independent limits:

1. The number of packets that are successfully encrypted and decrypted.

2. The number of packets that are *unsuccessfully* decrypted.

We have the former already.  The latter is what this issue tracks.  Once the limit is at risk of being exceeded, updating keys will force the attacker to start over.

Based on Felix's recommendation I am going to suggest that we limit forgery attempts to 2^36.  I have confirmed this with a close reading of the paper and concur.  There is a significant difference here between AES-GCM and ChaCha20+Poly1305 but we believe that setting a single limit is the best approach.  I will create a pull request to document this approach shortly.

## Notes

We could split the recommendation and allow a larger limit for AES-GCM, but I don't believe that 2^36 (~68 billion) is a significant barrier.  For reference, if you were able to saturate a 10Gbps link with 50 byte IP packets that contained plausible QUIC packets, and the recipient didn't treat this as DoS traffic, it would take 5 minutes.  If we assume that [Fastly's recent performance numbers](https://www.fastly.com/blog/measuring-quic-vs-tcp-computational-efficiency) are more likely representative of potential packet rates, then the limit goes up to 10s of hours.  I would point out that that volume of unproductive traffic is probably a good reason to use something other than a key update to deal with the sender.

A reminder that the underlying analysis assumes that packets are all 2^14 in size, which is the record size in TLS.  This is ~10 times larger than typical MTU on the Internet and many QUIC packets might be smaller than that, so - at least for AES-GCM - recommending a 2^36 limit is more conservative than necessary.  However, if we were to allow for different limits, we'd have to address the question of packet size, as a 2^16 byte packet will be used by some deployments.  For ChaCha20+Poly1305, the only value that matters is the number of packets, so we avoid that question.

Quoting from Felix's email for reference (with permission):

> To briefly recall, the analysis of the QUIC and DTLS 1.3 record layers in terms of robustness when handling maliciously injected packets we did [1] and I presented at QUIPS [2] suggests that a security degradation linear not only in regularly sent/received packets but *also* in the number of rejected packets should be taken into account. I.e., the unreliable-transport nature affects the integrity bounds, with unnoticeable, hence tolerated forgery attempts linearly degrading the underlying AEAD schemes' authenticity security level.
>
> I revisited the analysis done for TLS [3] together with Kenny, with the following conclusions (for both DTLS 1.3 and QUIC):
> 
> 1. AES-GCM
>    The (integrity) bounds in [3] for AES-GCM were already calculated with v = 2^60 forgery attempts ("verification attempts"). Therefore, the same data limits as for TLS 1.3 can be applied, but maybe a note should be added to the RFC that forgery attempts should be counted (and limited to < 2^60, if that's a practical concern).
>
> 2. ChaCha20+Poly1305
>    Integrity degradation for ChaCha20+Poly1305 is v * 2^-93 (for v verification/forgery attempts). In order to keep the same safety margin of 2^-57 as for TLS 1.3 [4], the number of overall received records (correctly verifying or rejected) must be limited to 2^36, requiring counting of received records and an according note in the RFC.
>
> We hope that these conclusions can answer the open questions on AEAD limits. Please let me know if any further investigation is needed, or in which way we can best contribute to adding the according notes to the draft standards.
>
> Best,
> Felix
>
> (with Christian, Kenny, Marc)
>
> [1] Marc Fischlin, Felix Günther, Christian Janson. "Robust Channels: Handling Unreliable Networks in the Record Layers of QUIC and DTLS". Preliminary version at http://felixguenther.info/Q20_RC.pdf
>
> [2] QUIPS slides on Robust Channels: https://github.com/chris-wood/NDSS20-QUIPS/blob/master/slides/quips2020_rch.pdf
>
> [3] Atul Luykx, Kenny Paterson. Limits on Authenticated Encryption Use in TLS". http://www.isg.rhul.ac.uk/~kp/TLS-AEbounds.pdf
>
> [4] https://tools.ietf.org/html/rfc8446#section-5.5

**Comment by martinthomson** [comment]
@ad-l Can I request that you review this analysis for correctness?  I know that you had reservations about the claims here.

**Comment by nibanks** [comment]
Another (simplier?) option could be to just update keys **much** more often. I've done performance tests with msquic on the effect of updating the key ~once per round trip. There was no noticeable effect on performance. In fact, more often than not, the key update perf tests happen to perform slightly better (within noise tolerance) than without key update.

I'm not suggesting we advocate once a round trip or anything near that; more like every 2^20 packets (or 2^30 bytes?) sent or received (including failed decryptions). As I understand it, this is an ultra conservative number as far as protecting from any kind of attacks, and should be _free_ as far as any performance impact on the connection. This also has the added benefit of exercising the key update scenario more often, which should hopefully improve interoperability of the feature.

**Comment by martinduke** [comment]
I support a change like this, but I am not qualified to say what the thresholds should be. There certainly isn't a reason to push it to very high limits. This change would not break interoperability, so it's a good one.

Banning CCM might break interoperability for someone, but I'm not too concerned, particularly if we can update all the browsers to not offer it.

A somewhat bigger concern is the lack of key update support in a lot of implementations: I read 9 of 15 servers as supporting it.

**Comment by martinthomson** [comment]
(Tentative) good news on CCM: Kenny identified a security analysis paper and it seems relatively simple to go from that to numbers.

I am not personally concerned about support for key update. 9/15 isn't 15, but that doesn't bother me. It turns out to be a little fiddly to implement, so if people decide to drop connections that hit these limits instead, the effect isn't that bad. Of course, if you want to follow Nick's suggestion, that works too (I remember Dan Bernstein seriously suggesting a cipher that updated every record, back when I first proposed this scheme for TLS).

**Comment by anrossi** [comment]
If we're looking for a new bound to initiate a key update, there's precedent in the SSH RFC to rekey after every 1 GB of data transferred. See [here](https://tools.ietf.org/html/rfc4253#section-9)

**Comment by martinduke** [comment]
I should point out that my "9 of 15" figure is based on the interop matrix, where the client initiates an early update and the server responds. That is a different thing than a server initiating an update when required by the spec, which fewer than 9 servers might do.

**Comment by martinthomson** [comment]
Based on [this paper analysing CCM](https://link.springer.com/content/pdf/10.1007%2F3-540-36492-7_7.pdf) by Jonsson, I have calculated the bounds on the number of forgeries and encryptions for CCM.  The numbers aren't great, but that probably doesn't matter much for the purposes to which CCM is generally put.

(I'm using the notation from the [AEBounds paper](https://www.isg.rhul.ac.uk/~kp/TLS-AEbounds.pdf) here because that is easier to type than some of what is being used in other papers.  For reference, q is the number of genuine encryptions, l is the length of each in 16 byte blocks, v is the number of forgeries, n is the key length in bits, t is the tag length in bits.)

## Confidentiality

Jonsson puts the advantage an attacker has over a generic PRF at `(ql)^2/2^(n+1)`.

If we assume the same numbers in the AEBounds paper for record/packet size, `l=2^10` (this is for consistency, most QUIC packets will be much smaller, in the order of 2^7, but we might also go as high as 2^12; this being dictated by the MTU).  For AES, `n=2^128`.

To keep the advantage an attacker has to 2^-60 (to match the analysis in TLS), we therefore need to keep q to below 2^24.5.  Somewhat unsurprisingly, that matches the numbers we have for AES-GCM.

## Integrity

Jonsson puts the advantage for an attacker over a generic PRF at `v/2^t + (2^l*(v + q))^2/2^(n+1)`.  As the first term is negligible even for large v (up to 2^64), we consider the second term alone and aim for a bound on the advantage of 2^-57 to match the analysis for other ciphers.

That leaves us with `v+q <= 2^26`.  As q is already established to be 2^24.5, we can say that v should be limited to 2^25 (or `log2(2^26 - 2^24.5)` if you want to be precise).

(As a side note, for t=64, as in AEAD_AES_128_CCM_8, the first term becomes relevant and our security bound limits the number of forgeries to 2^7, which is probably a bit limiting in practice.  That's good justification for not enabling AEAD_AES_128_CCM_8 by default; though different applications might use a different target bound than 2^-57.)

## Caveat

I'm an not a cryptographer.  These papers all read like chicken-chicken-chicken.  My calculation

**Comment by ad-l** [comment]
The shape of the bound seems plausible but I'll start a thread with the cryptographers to check everything

**Comment by fxguenther** [comment]
I reproduced Martin's analysis of CCM based on [Rogaway's 2011 evaluation](https://crossbowerbt.github.io/docs/crypto/rogaway_modes.pdf) of the bounds by [Jonsson](https://link.springer.com/content/pdf/10.1007%2F3-540-36492-7_7.pdf). (Myself not being an expert in producing such bounds, I found the second-level confirmation by Rogaway to be helpful.)

I essentially agree with the numbers, except that --when following the same metric as the [AEBounds paper](https://www.isg.rhul.ac.uk/~kp/TLS-AEbounds.pdf)-- the advantage over an ideal PRP (instead of PRF) should be considered. This simply means losing a 1/2 factor in the bound:

**Confidentiality bound:** `(ql)^2 / 2^n`

`q = 2^24.5` yields an advantage of `2^-59`. I'd hence say it's fine to keep TLS 1.3's recommendation on sending max. full-sized `2^24.5` records.


**Integrity bound:** `v/2^t + (l*(v + q))^2 / 2^n`

`v+q <= 2^25.5`, hence `v <= 24.5` yields an advantage of `2^-57` (matching the other analyses).


Most importantly, I concur with Martin that AEAD_AES_128_CCM_8 would need either much lower bounds or a different risk assessment: applying the same limit of `v = 24.5` would bring the advantage incurred through the `v/2^t` bound up to `2^-39.5`.


Would it still be the best approach to define the same anti-forgery limit across all cipher suites, even if this means going down to `2^24.5`?


Of course, I'm happy to discuss more with @ad-l !

**Comment by martinthomson** [comment]
Thanks for checking this @fxguenther.  I was a little concerned about the PRP/PRF split, so I'm glad to have you correct me.

I ultimately went with per-AEAD recommendations, and a lower limit for CCM.  That's mainly to establish some sort of uniformity around 2^-60 and 2^-57.  Even if those are basically arbitrary choices, at least I think they are defensible and using them uniformly establishes the right expectations about what the standard is.

I'm comfortable with specifying different limits for each AEAD in specifications.  In practice, however, I expect that a far lower tolerance for forgery attempts.

Assuming that these numbers work out, a review of the pull request would be greatly appreciated.  If there are other relevant papers, I'm always happy to pull those in as well.

**Comment by martinthomson** [comment]
I realize that I didn't respond to @anrossi.  That advice applies to the limits that TLS specifies already; this issue is primarily about responding to forgeries.

That said, 1G is pretty good advice.  However what we've seen from this analysis that 1G results in a lower safety margin than TLS aims for.  But it's close enough that if the AEAD is as good as the set we have here, you are probably OK.  However, this does depend a lot on the specific AEAD.  I'd be uncomfortable using so simple a guide in the general case.

**Comment by martinthomson** [comment]
@chris-wood points out an error in my calculation.  I had based my calculations on direct mapping of `l[E]` in the CCM analysis to `l` in our calculations.  That is, `l` was the number of blocks in the message.  But the analysis by Jonsson defines this as:

![image](https://user-images.githubusercontent.com/67641/81360611-b252b580-911f-11ea-8d81-d6824c685d51.png)

The definition of \Beta is the CCM function, so this function really reduces to 2 times the length of the message (in blocks), plus 1 (to account for the additional encryption).  We can ignore the extra 1 as that is absorbed by the tag length.

I'm updating the numbers in the PR.  The result is another halving of the number of packets.  @fxguenther, @ad-l, it's pretty clear that I'm out of my depth here, so your input would be highly valued.

**Comment by fxguenther** [comment]
Good catch, yes, the double passing over the message requires a factor 2 here. (The AD a may also be up to 3 blocks if I understand the header format correctly, but I guess that's negligible compared to the conservative 2^10 message size.)

**Comment by martinthomson** [comment]
Good point about the AD.  The AD - at least for QUIC - is always sent, so that fits in the same space.  We don't necessarily need to double count the tiny number of blocks, but as you say, counting that only once yields a negligible difference.  If we had a large AD that wasn't also transmitted, that would be different.

**Comment by fxguenther** [comment]
Agreed. To be precise: if I'm not mistaken the AD isn't counted at all right now (`l` accounts for the max. `2^10` payload blocks), but of course the difference is still negligible as long as the AD is small.
